### DISCUSSION TITLE
Response Cache-control header, based on requests made.

### DISCUSSION THREAD

**Discussion by TonyWeston** [quarkusio/quarkus#32071] [Q&A] [answered]
Hello.

During processing of a quarkus request, there are quarkus services that make other http requests to other servers.  These responses return 'Cache-Control' headers.

I want to set a quarkus response cache-control header, based on the received responses from the other servers - for example, if a server responds with JSON data, with a CacheControl: max-age= 5678, and my quarkus app has built a page based on that data, then I want my quarkus app to also include Cache-Control:max-age=5678 with the page.

The problem is that things are happening reactively, so the requests/responses from other servers is happening in another thread. The Qute template gets a CompletableFuture, and works fine.

BUT, I need something to trigger, once the qute templates have fully being populated, but before the response has returned back to the user.

Help!

[a comment by an automated account is not shown]

**Comment by gwenneg** [comment]
@geoand Do you know who could answer that question?

**Comment by geoand** [comment]
That would be me :)

I can't say I understand the relevance between the `CacheControl` header and template rendering.

Mind showing some example code (just pseudo-code is fine) showing what exactly you want to do?

**Reply by TonyWeston** [reply]
Scenario: need to build a page showing the current weather forecast, based on a weather API.
The API returns weather for a particular location, and responds with a JSON, with cache-control set to the time the next forecast will be updated.   I need to set my weather forecast page, to have a similar cache-control: max-age, so the cached browsers/cdn will know when to re-request the page.

Pseudocode:
@GET
public response buildWeatherPage() {
create a weatherTemplateInstance from a weatherTemplate
weatherTemplateInstance.data("forecast", weatherService.getForecast());

return Response.ok(weatherTemplateInstance).CacheControl(????).build();
}


The weatherService returns a completableFuture, which completes when the WeatherAPI returns with the required data.

The problem is, that the buildWeatherPage will return before the WeatherAPI has returned, and so its impossible to know how to set the CacheControl.

**Reply by TonyWeston** [reply]
ALSO...
I realise that one way of fixing this would be to block this buildWeatherPage method, waiting while the completableFuture completes,  but this would prevent other components in the page, which may themselves require data to be returned from other external servers, from starting to process.

**Comment by TonyWeston** [comment, marked as answer]
I think I have a solution that fits my usecase:

If I do:
 String responseHTML = weatherPageInstance.render();
 
 rather than returning the weatherPageInstance, this forces weatherPageInstance to fully render, which means all cache-controls, returned by requests for data to build the weatherPageInstance. will have resolved.
 
 Which means I can set the cache-control header in my controller. 
 
 :-)
 
 Not sure if this is the right, or best way,  but it works for me.

**Comment by geoand** [comment]
If the WeatherService returns a `CompletableFuture`, then you can make your REST Resource return `CompletableFuture<RestResponse<Whatever>>` and handle everything in an async manner
